perf/volumerhs.jl

```diff
@@ -0,0 +1,275 @@
+
+            s_G[i, j,  _ρ] = MJ * (ηx * fluxρ_x + ηy * fluxρ_y + ηz * fluxρ_z)
+            s_G[i, j,  _U] = MJ * (ηx * fluxU_x + ηy * fluxU_y + ηz * fluxU_z)
+            s_G[i, j,  _V] = MJ * (ηx * fluxV_x + ηy * fluxV_y + ηz * fluxV_z)
+            s_G[i, j,  _W] = MJ * (ηx * fluxW_x + ηy * fluxW_y + ηz * fluxW_z)
+            s_G[i, j,  _E] = MJ * (ηx * fluxE_x + ηy * fluxE_y + ηz * fluxE_z)
+
+            r_Hρ = MJ * (ζx * fluxρ_x + ζy * fluxρ_y + ζz * fluxρ_z)
+            r_HU = MJ * (ζx * fluxU_x + ζy * fluxU_y + ζz * fluxU_z)
+            r_HV = MJ * (ζx * fluxV_x + ζy * fluxV_y + ζz * fluxV_z)
+            r_HW = MJ * (ζx * fluxW_x + ζy * fluxW_y + ζz * fluxW_z)
+            r_HE = MJ * (ζx * fluxE_x + ζy * fluxE_y + ζz * fluxE_z)
+
+            # one shared access per 10 flops
+            for n = 1:Nq
+                Dkn = s_D[k, n]
+
+                r_rhsρ[n] += Dkn * r_Hρ
+                r_rhsU[n] += Dkn * r_HU
+                r_rhsV[n] += Dkn * r_HV
+                r_rhsW[n] += Dkn * r_HW
+                r_rhsE[n] += Dkn * r_HE
+            end
+
+            r_rhsW[k] -= MJ * ρ * gravity
+
+            sync_threads()
+
+            # loop of ξ-grid lines
+            @unroll for n = 1:Nq
+                Dni = s_D[n, i]
+                Dnj = s_D[n, j]
+
+                r_rhsρ[k] += Dni * s_F[n, j, _ρ]
+                r_rhsρ[k] += Dnj * s_G[i, n, _ρ]
+
+                r_rhsU[k] += Dni * s_F[n, j, _U]
+                r_rhsU[k] += Dnj * s_G[i, n, _U]
+
+                r_rhsV[k] += Dni * s_F[n, j, _V]
+                r_rhsV[k] += Dnj * s_G[i, n, _V]
+
+                r_rhsW[k] += Dni * s_F[n, j, _W]
+                r_rhsW[k] += Dnj * s_G[i, n, _W]
+
+                r_rhsE[k] += Dni * s_F[n, j, _E]
+                r_rhsE[k] += Dnj * s_G[i, n, _E]
+            end
+        end # k
+
+        @unroll for k in 1:Nq
+            MJI = vgeo[i, j, k, _MJI, e]
+
+            # Updates are a performance bottleneck
+            # primary source of stall_long_sb
+            rhs[i, j, k, _U, e] += MJI*r_rhsU[k]
+            rhs[i, j, k, _V, e] += MJI*r_rhsV[k]
+            rhs[i, j, k, _W, e] += MJI*r_rhsW[k]
+            rhs[i, j, k, _ρ, e] += MJI*r_rhsρ[k]
+            rhs[i, j, k, _E, e] += MJI*r_rhsE[k]
+        end
+    end
+    return
+end
+
+function main()
+    DFloat = Float32
+    nelem = 240_000
+
+    rng = StableRNG(123)
+
+    Nq = N + 1
+    nvar = _nstate + nmoist + ntrace
+
+    Q = 1 .+ CuArray(rand(rng, DFloat, Nq, Nq, Nq, nvar, nelem))
+    Q[:, :, :, _E, :] .+= 20
+    vgeo = CuArray(rand(rng, DFloat, Nq, Nq, Nq, _nvgeo, nelem))
+
+    # make sure the entries of the mass matrix satisfy the inverse relation
+    vgeo[:, :, :, _MJ, :] .+= 3
+    vgeo[:, :, :, _MJI, :] .= 1 ./ vgeo[:, :, :, _MJ, :]
+
+    D = CuArray(rand(rng, DFloat, Nq, Nq))
+
+    rhs = CuArray(zeros(DFloat, Nq, Nq, Nq, nvar, nelem))
+
+    threads=(N+1, N+1, N+1)
```

Comment:
```suggestion
    threads=(N+1, N+1)
```

Reply:
How did was that still in there... I apparently only fixed it on the AMD side.